core/lib/zksync_core/src/consensus/mod.rs

```diff
@@ -61,61 +60,155 @@ impl MainNodeConfig {
-    const RETRY_DELAY_INTERVAL: time::Duration = time::Duration::seconds(5);
-    loop {
-        match ctx.wait(client.fetch_l2_block_number()).await? {
-            Ok(head) => {
-                sync_state.set_main_node_block(head);
-                ctx.sleep(DELAY_INTERVAL).await?;
+pub type FetcherConfig = executor::Config;
+
+/// External node consensus config.
+#[derive(Debug)]
+pub struct Fetcher {
+    pub config: FetcherConfig,
+    pub sync_state: SyncState,
+    pub client: Box<dyn MainNodeClient>,
+}
+
+impl Fetcher {
+    /// Periodically fetches the head of the main node
+    /// and updates `SyncState` accordingly.
+    async fn fetch_state_loop(&self, ctx: &ctx::Ctx) -> ctx::Result<()> {
+        const DELAY_INTERVAL: time::Duration = time::Duration::milliseconds(500);
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            match ctx.wait(self.client.fetch_l2_block_number()).await? {
+                Ok(head) => {
+                    self.sync_state.set_main_node_block(head);
+                    ctx.sleep(DELAY_INTERVAL).await?;
+                }
+                Err(err) => {
+                    tracing::warn!("main_node_client.fetch_l2_block_number(): {err}");
+                    ctx.sleep(RETRY_INTERVAL).await?;
+                }
             }
-            Err(err) => {
-                tracing::warn!("main_node_client.fetch_l2_block_number(): {err}");
-                ctx.sleep(RETRY_DELAY_INTERVAL).await?;
+        }
+    }
+
+    /// Fetches genesis from the main node.
+    async fn fetch_genesis(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::Genesis> {
+        let genesis = ctx
+            .wait(self.client.fetch_consensus_genesis())
+            .await?
+            .context("fetch_consensus_genesis()")?
+            .context("main node is not running consensus component")?;
+        Ok(zksync_protobuf::serde::deserialize(&genesis.0).context("deserialize(genesis)")?)
+    }
+
+    /// Fetches (with retries) the given block from the main node.
+    async fn fetch_block(&self, ctx: &ctx::Ctx, n: MiniblockNumber) -> ctx::Result<FetchedBlock> {
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            let res = ctx.wait(self.client.fetch_l2_block(n, true)).await?;
+            match res {
+                Ok(Some(block)) => return Ok(block.try_into()?),
+                Ok(None) => {}
+                Err(err) if is_transient(&err) => {}
+                Err(err) => {
+                    return Err(anyhow::format_err!("client.fetch_l2_block({}): {err}", n).into());
+                }
             }
+            ctx.sleep(RETRY_INTERVAL).await?;
         }
     }
-}
 
-/// External node consensus config.
-#[derive(Debug, Clone)]
-pub struct FetcherConfig {
-    pub executor: executor::Config,
-}
+    /// Fetches blocks from the main node in range `[cursor.next()..end)`.
+    async fn fetch_blocks(
```

Comment:
It looks weird that EN now has two JSON-RPC-based fetchers. Could [the existing fetcher](https://github.com/matter-labs/zksync-era/blob/10e3a3ec50c84b3533af111d532e0b528c0b4966/core/lib/zksync_core/src/sync_layer/fetcher.rs#L193) be adapted to be used here, or, if it's bad, could it be entirely replaced with this one?

Reply:
deduplicated